keyforge-refactor.ino

```diff
@@ -330,50 +326,61 @@ class GameVisuals {
 };
 
-GameVisuals visuals;
+class MachineState {
+  public:
+
+  GameState game;
+  const GameVisuals visuals;
 
-int lastButton;
-unsigned long lastRender = 0;
+  int lastButton;
+  unsigned long lastRender = 0;
 
-void loop() {
-  const int currentButtons = lcd.readButtons();
+  void handleGame() {
+    const int currentButtons = lcd.readButtons();
 
-  if (currentButtons != lastButton) {
-    if (currentButtons == BUTTON_SELECT) {
-      lcd.clear();
-      game.nextPhase();
-    }
-    if (currentButtons == BUTTON_LEFT) {
-      game.currentStat--;
-      if (game.currentStat < 0) {
-        game.currentStat = 2;
+    if (currentButtons != lastButton) {
+      if (currentButtons == BUTTON_SELECT) {
+        lcd.clear();
+        game.nextPhase();
+      }
+      if (currentButtons == BUTTON_LEFT) {
+        game.currentStat--;
+        if (game.currentStat < 0) {
+          game.currentStat = 2;
+        }
+      }
+      if (currentButtons == BUTTON_RIGHT) {
+        game.currentStat++;
+        game.currentStat %= 3;
+      }
+      if (currentButtons == BUTTON_UP) {
+        if (game.currentPhase == mainPlayPhase) game.changePlayerStat(1);
+        if (game.currentPhase == chainsPrompt) game.changeHandSize(1);
+        if (game.currentPhase == forgePrompt) game.changeForgeMod(1);
+      }
+      if (currentButtons == BUTTON_DOWN) {
+        if (game.currentPhase == mainPlayPhase) game.changePlayerStat(-1);
+        if (game.currentPhase == chainsPrompt) game.changeHandSize(-1);
+        if (game.currentPhase == forgePrompt) game.changeForgeMod(-1);
+      }
+      // toggle between players without finish the current turn:
+      if (currentButtons == BUTTON_RIGHT + BUTTON_LEFT) {
+        game.currentPlayer = game.currentPlayer == 1 ? 2 : 1;
       }
     }
-    if (currentButtons == BUTTON_RIGHT) {
-      game.currentStat++;
-      game.currentStat %= 3;
-    }
-    if (currentButtons == BUTTON_UP) {
-      if (game.currentPhase == mainPlayPhase) game.changePlayerStat(1);
-      if (game.currentPhase == chainsPrompt) game.changeHandSize(1);
-      if (game.currentPhase == forgePrompt) game.changeForgeMod(1);
-    }
-    if (currentButtons == BUTTON_DOWN) {
-      if (game.currentPhase == mainPlayPhase) game.changePlayerStat(-1);
-      if (game.currentPhase == chainsPrompt) game.changeHandSize(-1);
-      if (game.currentPhase == forgePrompt) game.changeForgeMod(-1);
-    }
-    // toggle between players:
-    if (currentButtons == BUTTON_RIGHT + BUTTON_LEFT) {
-      game.currentPlayer = game.currentPlayer == 1 ? 2 : 1;
+
+    lastButton = currentButtons;
+    unsigned long currentTime = millis();
+
+    if (currentTime - lastRender > 250) {
+      lastRender = currentTime;
+      visuals.render(game);
     }
   }
+};
 
-  lastButton = currentButtons;
-  unsigned long currentTime = millis();
+MachineState machine;
 
-  if (currentTime - lastRender > 250) {
-    lastRender = currentTime;
-    visuals.render(game);
-  }
+void loop() {
```

Comment:
I'd break down the buttons handling and display rendering into two methods for `handleRender` and `handleButtons`

Reply:
- [ ] Split handleGame() logic